services/data_sources/json_notification_data_source.rb

```diff
@@ -0,0 +1,162 @@
+require "securerandom"
+require_relative "notification_data_source"
+require_relative "../../shared/logging_module"
+require_relative "../../shared/json_convertible"
+require_relative "../../shared/models/notification"
+
+module FastlaneCI
+  # Mixin the JSONConvertible class for Notification
+  class Notification
+    include FastlaneCI::JSONConvertible
+  end
+
+  # Data source for notifications backed by JSON
+  class JSONNotificationDataSource < NotificationDataSource
+    include FastlaneCI::JSONDataSource
+    include FastlaneCI::Logging
+
+    class << self
+      attr_accessor :file_semaphore
+    end
+
+    # @return [FastlaneCI::GitRepo]
+    attr_accessor :git_repo
+
+    # Can't have us reading and writing to a file at the same time
+    JSONNotificationDataSource.file_semaphore = Mutex.new
+
+    # Reloads notifications from the notifications data source after instantiation
+    #
+    # @param  [Any] params
+    def after_creation(**params)
+      if params.nil?
+        raise "Either user or a provider credential is mandatory."
+      else
+        if !params[:user] && !params[:provider_credential]
+          raise "Either user or a provider credential is mandatory."
+        else
+          params[:provider_credential] ||= params[:user].provider_credential(type: ProviderCredential::PROVIDER_CREDENTIAL_TYPES[:github])
+          @git_repo = FastlaneCI::GitRepo.new(git_config: self.json_folder_path, provider_credential: params[:provider_credential])
+        end
+      end
+
+      reload_notifications
+    end
+
+    # Returns an array of notifications from the notifications JSON file stored
+    # in the configuration git repo
+    #
+    # @return [Array[Notification]]
+    def notifications
+      JSONProjectDataSource.projects_file_semaphore.synchronize do
+        path = git_repo.file_path("notifications.json")
+        return [] unless File.exist?(path)
+
+        return JSON.parse(File.read(path)).map(&Notification.method(:from_json!))
+      end
+    end
+
+    # Writes the notifications array to the git repo as JSON and commit them as
+    # the CI user
+    #
+    # @param  [Array[Notification]] notifications
+    def notifications=(notifications)
```

Comment:
Small API nitpick: As an API consumer of this class, I would be surprised that calling a setter method results in an IO operation. I would suggest an alternative of having a `persist_notifications` (or something like that) private method that can be called at the end of methods such as `update_notification!`. I would feel more idiomatic ruby to have `notification` getter and setter operate on the `@notification` instance variable. Just a suggestion, tho!

Reply:
This is a great point. We might want to consider putting together a best practices guide for this sort of thing because I've actually run into it a couple of times.

Reply:
Ahh, yeah. I tried to keep it consistent with the code we currently have in the `json_project_data_source.rb` [1](https://github.com/fastlane/ci/blob/master/services/config_data_sources/json_project_data_source.rb#L84-L94), [2](https://github.com/fastlane/ci/blob/master/services/config_data_sources/json_project_data_source.rb#L116-L120).

That being said, I am a fan of the private `persist_notifications` method you're suggesting. I agree it feels more like idiomatic ruby.

There's a lot of shared logic in both the `json_project_data_source.rb` and `json_notification_data_source.rb`, so I think it might also be beneficial to abstract the shared logic in both of those classes to an abstract base class. However, the thing keeping me from making those changes is that it might be a premature abstraction, or might be suited for another PR.

Thoughts?

Reply:
I would put that in another PR, and keep the focus of this one 💯